src/plic.cpp

```diff
@@ -0,0 +1,234 @@
+    (void) context;
+#ifdef DEBUG_PLIC_MMIO
+    (void) fprintf(stderr, "plic: mmio read offset=0x%lx log2_size=%d\n", (long) offset, log2_size);
+#endif
+
+    // Our PLIC only supports aligned 32-bit reads
+    if (offset & 3 || log2_size != 2 || offset > PMA_PLIC_LENGTH) {
+        return false;
+    }
+
+    switch (offset) {
+        case plic_csr_rel_addr::priority1:
+        case plic_csr_rel_addr::priority2:
+        case plic_csr_rel_addr::priority3:
+        case plic_csr_rel_addr::priority4:
+        case plic_csr_rel_addr::priority5:
+        case plic_csr_rel_addr::priority6:
+        case plic_csr_rel_addr::priority7:
+        case plic_csr_rel_addr::priority8:
+        case plic_csr_rel_addr::priority9:
+        case plic_csr_rel_addr::priority10:
+        case plic_csr_rel_addr::priority11:
+        case plic_csr_rel_addr::priority12:
+        case plic_csr_rel_addr::priority13:
+        case plic_csr_rel_addr::priority14:
+        case plic_csr_rel_addr::priority15:
+        case plic_csr_rel_addr::priority16:
+        case plic_csr_rel_addr::priority17:
+        case plic_csr_rel_addr::priority18:
+        case plic_csr_rel_addr::priority19:
+        case plic_csr_rel_addr::priority20:
+        case plic_csr_rel_addr::priority21:
+        case plic_csr_rel_addr::priority22:
+        case plic_csr_rel_addr::priority23:
+        case plic_csr_rel_addr::priority24:
+        case plic_csr_rel_addr::priority25:
+        case plic_csr_rel_addr::priority26:
+        case plic_csr_rel_addr::priority27:
+        case plic_csr_rel_addr::priority28:
+        case plic_csr_rel_addr::priority29:
+        case plic_csr_rel_addr::priority30:
+        case plic_csr_rel_addr::priority31:
+            // A valid implementation can hardwire all input priority levels.
+            // We hardwire all supported interrupt sources to the lowest priority
+            *val = PLIC_LOWEST_IRQ_PRIORITY;
+            return true;
+        case plic_csr_rel_addr::pending:
+            *val = plic_read_pending(a);
+            return true;
+        case plic_csr_rel_addr::enabled:
+            // A valid implementation can hardwire interrupt routing to a fixed hart context.
+            // We hardwire all supported interrupt source to be always enabled in context 0.
+            *val = PLIC_ENABLED_IRQ_MASK;
+            return true;
+        case plic_csr_rel_addr::claim_complete:
+            return plic_read_claim_complete(a, val);
+        default:
+            // Other PLIC CSRs are WARL hardwired to 0
+            *val = 0;
+            return true;
+    }
+}
+
+/// \brief PLIC device read callback. See ::pma_write.
+static execute_status plic_write(void *context, i_device_state_access *a, uint64_t offset, uint64_t val,
+    int log2_size) {
+    (void) context;
+#ifdef DEBUG_PLIC_MMIO
+    (void) fprintf(stderr, "plic: mmio write offset=0x%lx log2_size=%d val=0x%x\n", (long) offset, log2_size,
+        (int) val);
+#endif
+
+#if 0
```

Comment:
Have we stopped using BBL already?

Reply:
Yes, right? And we won't support I a guess?

Reply:
It's pretty obsolete, right? Should we just sunset it?